core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,207 @@
+  protected interface Sink {
+    /**
+     * Sends response headers to the remote end points.
+     *
+     * @param headers the headers to be sent to client.
+     */
+    void writeHeaders(Metadata headers);
 
-  protected AbstractServerStream(WritableBufferAllocator bufferAllocator,
-                                 int maxMessageSize) {
-    super(bufferAllocator, maxMessageSize);
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent.
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean flush);
+
+    /**
+     * Sends trailers to the remote end point. This call implies end of stream.
+     *
+     * @param trailers metadata to be sent to end point
+     * @param headersSent {@code true} if response headers have already been sent.
+     */
+    void writeTrailers(Metadata trailers, boolean headersSent);
+
+    void request(int numMessages);
+
+    void cancel(Status status);
   }
 
-  /**
-   * Sets the listener to receive notifications. Must be called in the context of the transport
-   * thread.
-   */
-  public final void setListener(ServerStreamListener listener) {
-    this.listener = checkNotNull(listener);
+  private final MessageFramer framer;
+  private boolean outboundClosed;
+  private boolean headersSent;
 
-    // Now that the stream has actually been initialized, call the listener's onReady callback if
-    // appropriate.
-    onStreamAllocated();
+  protected AbstractServerStream(WritableBufferAllocator bufferAllocator) {
+    this.framer = new MessageFramer(this, bufferAllocator);
   }
 
   @Override
-  protected ServerStreamListener listener() {
-    return listener;
+  protected abstract TransportState transportState();
+
+  protected abstract Sink abstractServerStreamSink();
+
+  @Override
+  protected MessageFramer framer() {
+    return framer;
   }
 
   @Override
-  protected void receiveMessage(InputStream is) {
-    inboundPhase(Phase.MESSAGE);
-    listener().messageRead(is);
+  public final void request(int numMessages) {
+    abstractServerStreamSink().request(numMessages);
   }
 
   @Override
   public final void writeHeaders(Metadata headers) {
     Preconditions.checkNotNull(headers, "headers");
 
-    outboundPhase(Phase.HEADERS);
     headersSent = true;
-    internalSendHeaders(headers);
-    outboundPhase(Phase.MESSAGE);
+    abstractServerStreamSink().writeHeaders(headers);
   }
 
   @Override
-  public final void writeMessage(InputStream message) {
-    if (outboundPhase() != Phase.MESSAGE) {
-      throw new IllegalStateException("Messages are only permitted after headers and before close");
-    }
-    super.writeMessage(message);
+  public final void deliverFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
+    abstractServerStreamSink().writeFrame(frame, endOfStream ? false : flush);
```

Comment:
`endOfStream ? false : flush` reads a bit easier for me, but it's also what was in the preexisting code.